### ISSUE TITLE
RLS docs encourage using a vulnerable RLS policy

### ISSUE THREAD

**Issue by aantn** [supabase/supabase#21391]
The docs make it clear that the `user_metadata` inside JWTs is not to be trusted. See [here](https://supabase.com/docs/guides/auth/row-level-security#authjwt:~:text=user_metadata%20%2D%20can%20be%20updated%20by%20the%20authenticated%20user%20using%20the%20supabase.auth.update()%20function.%20It%20is%20not%20a%20good%20place%20to%20store%20authorization%20data.) and [here](https://github.com/supabase-community/supabase-custom-claims?tab=readme-ov-file#whats-the-difference-between-authusersraw_app_meta_data-and-authusersraw_user_meta_data).

However, the RLS docs explicitly recommend relying on `user_metadata` inside JWTs. See [here](https://supabase.com/docs/guides/auth/sso/auth-sso-saml#row-level-security:~:text=auth.jwt()%23%3E%3E%27%7Buser_metadata%2Ciss%7D%27%20Returns%20the%20identity%20provider%27s%20SAML%202.0%20EntityID)

To add insult to injury, when logging in with SSO, the user's actual custom_claims and groups are written to auth.users.raw_user_meta_data which is certainly not secure. (It can be overriden easily.) I assume that should be ignored entirely.

I have two questions:

1. Is the recommended RLS policy indeed vulnerable to privilege escalation. Or are the docs wrong and if so, which page should be trusted?
2. What is the secure way to check a user's SSO claims and/or groups? Is it via `auth.jwt()`? Is it via the `auth.identities` table as [implied here](https://supabase.com/docs/guides/auth/sso/auth-sso-saml#row-level-security:~:text=Accessing%20the%20stored%20attributes)? Or some other manner?

**Comment by aantn** [comment]
I've been testing and I'm nearly certain now that the [RLS docs (for SSO)](https://supabase.com/docs/guides/auth/sso/auth-sso-saml#row-level-security) are wrong and dangerous.

E.g. a malicious user can bypass the suggested RLS policy by changing their ISS in javascript:

```
await this.client.auth.update({ data: { iss: "https://lets.be.someone.else.today.com" }});
```

That gets reflected in `auth.jwt` in SQL meaning the docs suggestion that you can rely on the ISS value is wrong. Likewise, you cannot rely on custom claims or groups as they appear in the JWT.

**Comment by aantn** [comment]
I'm starting to suspect that this is a bug and these values were intended to be mapped to `raw_app_meta_data` not `raw_auth_users`

For example, from [the SSO docs](https://supabase.com/docs/guides/auth/sso/auth-sso-saml#understanding-attribute-mappings):
![Screenshot 2024-02-20 at 14 36 56](https://github.com/supabase/supabase/assets/494087/8a393720-9226-4081-a74d-e68694c62089)

I haven't tested that flow, but if these values were intended to be in `raw_app_meta_data` then everything starts to make more sense. (In which case this is a real security bug, not just a docs issue.)

**Comment by divmgl** [comment]
The point of `user_metadata` is that it can be updated by the client.

It seems like it's a documentation bug.

**Comment by aantn** [comment]
@divmgl is there a secure way to check claims and groups from SSO?

The data seems to be reflected in auth.identities, but there are no clear docs saying you can read it from there safely.

**Comment by encima** [comment]
This does seem like a docs bug, thanks for calling it out @aantn. `raw_app_meta_data` is indeed the property to use there and there is an example implementation [here](https://github.com/supabase-community/supabase-custom-claims)

**Comment by aantn** [comment]
@encima `raw_app_meta_data` doesn't work for this. The code you shared allows you to add your own claims to `raw_app_meta_data`. It doesn't let you access existing claims that exist on the SSO provider , such as group membership.

Existing data from the SSO provider is copied to `raw_user_meta_data` and a section of the JWT that can be controlled by the end user.

What is the recommended way to access verified data that already exists on the SSO provider?

**Comment by encima** [comment]
Hi @aantn ,

That is still possible. See the docs  just above the link (in Accessing the stored attributes) [here](https://supabase.com/docs/guides/auth/sso/auth-sso-saml#remove-a-connection). Mapped attributes are stored in the app_metadata (as well as the `identities` table).

**Comment by aantn** [comment]
@encima I assume you're referring to `app_metadata` in the JWT, correct?

If so, the relevant data is **not** there. It is in `user_metadata` instead which is insecure. Here is a real (censored) JWT from a managed supabase instance:

```
{
    "aal": "aal1",
    "amr": [
        {
            "method": "password",
            "timestamp": <snip>
        }
    ],
    "aud": "authenticated",
    "exp": <snip>,
    "iat": <snip>,
    "iss": "<snip>",
    "sub": "<snip>",
    "role": "authenticated",
    "email": "<snip>",
    "phone": null,
    "session_id": "<snip>",
    "app_metadata": {
        "provider": "google",
        "providers": [
            "google",
            "azure"
        ]
    },
    "user_metadata": {
        "iss": "https://login.microsoftonline.com/<snip>",
        "sub": "<snip>",
        "name": "<snip>",
        "email": "<snip>",
        "picture": "https://lh3.googleusercontent.com/a/<snip>",
        "full_name": "<snip>",
        "avatar_url": "https://lh3.googleusercontent.com/a/<snip>",
        "provider_id": "<snip>",
        "custom_claims": {
            "tid": "<snip>",
            "wids": [
                <snip>
            ],
            "email": "<snip>",
            "groups": [
                "<snip>"            ]
        },
        "email_verified": true,
        "phone_verified": false,
    }
}
```

There is no secure copy of `custom_claims` or `groups` in the JWT. The only copy in the JWT is user-controllable.

**Comment by encima** [comment]
@aantn You may be right here.

To confirm, are you asking for the `groups` scope when logging in the user (for azure and/or google)?
Using Microsoft as the example, how have you configured the claims to be emitted (see [here](https://learn.microsoft.com/en-us/entra/identity/hybrid/connect/how-to-connect-fed-group-claims))?

How does your [mapping](https://learn.microsoft.com/en-us/entra/identity/app-provisioning/customize-application-attributes) look? And for [Supabase](https://supabase.com/docs/guides/auth/sso/auth-sso-saml#understanding-attribute-mappings)?

When that is configured for the Auth, you will see it in the `identities` table.

**Comment by aantn** [comment]
@encima, sorry, I should have clarified. We are using OAuth not SAML, but does that matter in this context? (We configured the claims in the OIDC app in Azure, as you can see below.)
 
![Screenshot 2024-02-21 at 12 46 07 PM](https://github.com/supabase/supabase/assets/494087/9b77f08e-1921-4280-8b97-4b7b2ab83527)

We really only have one major concern and question: is it safe to rely on the `identities` table. If so, we are good to go. We just want to verify that the `identities` table is **not** user controlled like parts of the JWT are.

**Comment by encima** [comment]
@aantn It sort of matters:
- For OAuth, you only need to make sure it is in your `scopes`
- For SAML, it is more configurable as you can add/edit mappings so the response is more flexible.

The `auth` schema itself is protected and managed by Supabase, and this is also true for the `identities` table within it, yes.

**Comment by aantn** [comment]
@encima however you configure it, wont it still get written user_metadata in the JWT and auth.users table? So it doesn't matter whether we customize it and map the names, they're still written somewhere insecure.

> The auth schema itself is protected and managed by Supabase, and this is also true for the identities table within it, yes.
Sorry for being a nag, but I want to be 100% certain about this, as `auth.jwt()` and `auth.users.raw_user_meta_data` are also in the auth schema, yet parts of them are user-controllable. **Is a user able to control the groups and claims written to the identities table or is this data guaranteed to come only from the SSO provider?**

**Comment by encima** [comment]
@aantn You are not being a nag, at all! You have raised very good points and this is not an area where you want to operate on guesswork.
I can confirm that the user has no way to update the `identities` table

**Comment by aantn** [comment]
Thank you, perfect.

**Comment by encima** [comment]
@aantn glad this solves your problem! For the docs, how would you like to see them improved:
- clarification on identities table?
- clearer information about the client_metadata vs app_metadata?
- all of the above?
- something else

**Comment by hf** [comment]
Auth team is looking into this.

**Comment by aantn** [comment]
Yes, all of the above, with explicit warnings "DO NOT USE THE raw_user_meta_data IN RLS" and "DO NOT USE user_metadata IN THE JWT IN YOUR POLICIES"

**Comment by Hallidayo** [comment]
Hi everyone, A a pull request has been merged that fixes the issue I'm going to close this one now.
